packages/api/src/services/actions/global.ts

```diff
@@ -1,71 +1,136 @@
 import { BigNumber } from "ethers";
 import { nowS } from "../../libs/utils";
+import { Handlers as EmpActions } from "./emp";
+import { Handlers as LspActions } from "./lsp";
 
 // actions use all the app state
 type Dependencies = AppState;
 type Config = undefined;
 
 export function Handlers(config: Config, appState: Dependencies): Actions {
   const { stats } = appState;
-  const queries = [Queries.Emp(appState), Queries.Lsp(appState)];
+  const contractActions = [EmpActions(config, appState), LspActions(config, appState)];
 
   const actions: Actions = {
     echo(...args: Json[]) {
       return args;
     },
     async listActive() {
       return bluebird.reduce(
-        queries,
-        async (result: AllContractStates[], query) => {
-          return result.concat(await query.listActive());
+        contractActions,
+        async (result: AllContractStates[], actions) => {
+          return result.concat((await actions.listActive()) as AllContractStates[]);
         },
         []
       );
     },
     async listExpired() {
       return bluebird.reduce(
-        queries,
-        async (result: AllContractStates[], query) => {
-          return result.concat(await query.listExpired());
+        contractActions,
+        async (result: AllContractStates[], actions) => {
+          return result.concat((await actions.listExpired()) as AllContractStates[]);
         },
         []
       );
     },
     // return the first address found
     async getState(address: string) {
       assert(address, "requires a contract address");
-      for (const query of queries) {
-        try {
-          const state = await query.getAny(address);
-          return query.getFullState(state);
-        } catch (err) {
-          // do nothing
+      for (const action of contractActions) {
+        if (await action.hasAddress(address)) {
+          return action.getState(address);
         }
       }
       throw new Error("Unable to find contract address " + address);
     },
-    async tvl(currency: CurrencySymbol = "usd") {
+    async globalTvl(currency: CurrencySymbol = "usd") {
       const sum = await bluebird.reduce(
-        queries,
-        async (sum, queries) => {
-          return sum.add((await queries.getTotalTvl(currency)) || "0");
+        contractActions,
+        async (sum, actions) => {
+          return sum.add(((await actions.tvl(undefined, currency)) as string) || "0");
         },
         BigNumber.from("0")
       );
       return sum.toString();
     },
-    async tvlHistoryBetween(start = 0, end: number = nowS(), currency: CurrencySymbol = "usd") {
+    async tvl(address: string, currency: CurrencySymbol = "usd") {
+      assert(address, "requires a contract address");
+      for (const action of contractActions) {
+        if (await action.hasAddress(address)) {
+          return action.tvl([address], currency);
+        }
+      }
+      throw new Error("Unable to find TVL for address " + address);
```

Comment:
nit: shouldn't this just be that the address wasn't found -- it's unrelated to the action type, right?

Reply:
Technically you are right but I like that there is some context in the error message since you could show this directly to user and they would understand that the app was trying to get tvl for an address.  If the error was just "address not found" for all the calls, its a bit more cryptic and will make it hard to immediately know the call that's failing.

Reply:
if you feel strongly on this though lets discuss, as it will cause many of the global errors to change